### PR TITLE
Proof of concept for `context` idea

### PR DESCRIPTION
**Pull request by acstll**
This is an implementation of that discussed in here https://github.com/picodom/picodom/issues/26

It's very simple. Changes are:

- exports a new function called `setContext`
- all Components receive a third argument `(data, children, context)`

This third argument will be `null` in case `setContext` was never called, or it will be whatever you passed to `setContext`.

Tests have been added to keep coverage at 100%.

The use case is the same as explained in the [React documentation for context](https://facebook.github.io/react/docs/context.html):

> In some cases, you want to pass data through the component tree without having to pass the props down manually at every level.

Of course, this here is waaaaay more simple than context is in React.

If you like this, we can discuss a bit the naming (I don't really like `context` and would prefer something else) and merge it, since it's a very tiny change.

Otherwise, there's also the option of making it a separate package, like `picodom-context`, which would only export `{ h, setContext }`.

What do you think?

### CODE REVIEW

**Comment by jbucaran** [comment]
@acstll Thank you for this excellent POC. 

I'd prefer a separate package for the following reasons:

- **Simplicity**: We can't just add everything on top of picodom, but we can this time. This feature can perfectly live outside core.
- **Maintenance**: Without this feature there are less tests and code to maintain in the main repo.
- **Flexibility**: "Of course, this here is waaaaay more simple than context is in React." As a third party library, different authors can add just the right amount of "way" they want.

**Comment by jbucaran** [comment]
@acstll On the other hand, this is a very small addition, so let's discuss. I can change my mind.

An official package is another option too! 🤔

**Comment by jbucaran** [comment]
Comments? 🙏 @zaceno @andyrj @ngryman

**Comment by acstll** [comment]
You're very welcome. My opinion is also divided. It's such a small addition it wouldn't hurt, but it also makes sense as a separate package, so you use it only if you really need it.

Yep, comments would be highly appreciated.

**Comment by zaceno** [comment]
I think, unless we're sure this is something most users of picodom will want, this would be better as a separate package. That doesn't mean I'm against *ever* adding it. I think we need to see some real world usage first.

WOuld it be possible to have a working demonstration of how this is similar, but simpler than, context in React? 🙏

File: src/h.js [hunk 1]
```diff
@@ -1,5 +1,10 @@
 var i
 var stack = []
+var context = null
```
Comment (jbucaran) [inline src/h.js]: @acstll This means all users of `h` share the same context. Is that desirable?

Comment (acstll) [reply src/h.js]: I think it is, that's the idea. Having access to very common stuff. When you say "users" you mean any module (internal or external) that imports `h`, right?

**Comment by acstll** [comment]
> I think we need to see some real world usage first.

@zaceno I pretty much agree with this. I'm in favor of a separate package. If it turns out to be a good idea, we can evaluate adding it to core, but only then. I can create the package, and it could be later added to the picodom organization if you feel that's the right thing.

Also, being that `context` is a shared object (see https://github.com/picodom/picodom/pull/27#pullrequestreview-61438265), it makes a lot of sense for it to be separate.

Do you have any ideas for a better name, instead of `context`?

> WOuld it be possible to have a working demonstration of how this is similar, but simpler than, context in React? 

it's simpler because you don't need to declare anything, you just get it as an argument (like in Preact). It's the same because it's just an object you can use, not related to props or state (in picodom there's no such thing as state).

I "needed" this because I want to try and build something like (but smaller) `react-router` 4, which is routing using components. I can try and do that, so it'll do as a working demontration which is also useful.

**Comment by zaceno** [comment]
@acstll 

> I want to try and build something like (but smaller) react-router 4, which is routing using components. I can try and do that, so it'll do as a working demontration which is also useful.

Please do - that'd be great! I haven't used React myself, so I only have a very vague idea of what "context" is in that context. So I'm very curious to see how this construct would be used in practice.

I'm imagining things like you could attach the call to rerender the dom onto the context, perhaps, or to rerender certain parts of the vdom only. Or state management solutions. My imagination runs so far ahead of me that I need someone to show me some nuts and bolts :)

File: src/h.js (hunk 1 above)
Comment (jbucaran) [reply src/h.js]: Yes, that's what I meant.

**Comment by jbucaran** [comment]
Another issue about adding this to core that I didn't address [here](https://github.com/picodom/picodom/pull/27#issuecomment-328019824) is that it would create a precedent: where do we draw the line of what's core and what's not core?

**Comment by jbucaran** [comment]
> @acstll I can create the package, and it could be later added to the picodom organization if you feel that's the right thing. 

Sounds good to me.

> Do you have any ideas for a better name, instead of context?

Here are some ideas:

- `context`: Easier to remember since it's based on the same concept as [React's context](https://facebook.github.io/react/docs/context.html).
- `sharedProps`
- `wire`
- `channel`
- `medium`

**Comment by acstll** [comment]
> I can create the package

I'll do that then, thanks a lot!

**Comment by acstll** [comment]
Feel free to close this :)

**Comment by jbucaran** [comment]
Let's do it then! 💪

**Comment by jbucaran** [comment]
P.S: What did you decide to call it?

**Comment by acstll** [comment]
> What did you decide to call it?

I think I'll go with `context`, if you come from React or Preact you know exactly what it means.

**Comment by jbucaran** [comment]
@acstll Sounds good.

**Comment by ngryman** [comment]
@jbucaran Hey, so sorry for the flatline heartbeat lately, so much stuff on my side 😜.

@acstll I think this is a cool proposal but it shouldn't be part of the core. `picodom` is too low-level for this. The notion of components does not really exist here, it's an additional layer left to the user.

**Pull request closed without merging**